zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+                for _ in calldata_count..MAX_CALLDATA {
+                    config.assign_tx_row(
+                        &mut region,
+                        offset,
+                        0, // tx_id
+                        TxFieldTag::CallData,
+                        0,
+                        F::zero(),
+                        &mut raw_pi_vals,
+                    )?;
+                    offset += 1;
+                }
+
+                // rpi_rlc and rand_rpi cols
+                let (rpi_rand, rpi_rlc) =
+                    config.assign_rlc_pi(&mut region, self.rand_rpi, raw_pi_vals)?;
+
+                Ok(vec![
+                    rpi_rand,
+                    rpi_rlc,
+                    chain_id,
+                    state_root,
+                    prev_block_hash,
+                ])
+            },
+        )?;
+
+        // Constrain raw_public_input cells to public inputs
+        for (i, pi_cell) in pi_cells.iter().enumerate() {
+            layouter.constrain_instance(pi_cell.cell(), config.pi, i)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod pi_circuit_test {
+    use super::*;
+
+    use ethers_core::{
+        types::{NameOrAddress, TransactionRequest},
+        utils::keccak256,
+    };
+    use ethers_signers::{LocalWallet, Signer};
+    use halo2_proofs::{
+        dev::{MockProver, VerifyFailure},
+        pairing::bn256::Fr,
+    };
+    use pretty_assertions::assert_eq;
+    use rand::{CryptoRng, Rng, SeedableRng};
+    use rand_chacha::ChaCha20Rng;
+
+    // Copied from tx_circuit
+    fn rand_tx<R: Rng + CryptoRng>(mut rng: R, chain_id: u64) -> Transaction {
+        let wallet0 = LocalWallet::new(&mut rng).with_chain_id(chain_id);
+        let wallet1 = LocalWallet::new(&mut rng).with_chain_id(chain_id);
+        let from = wallet0.address();
+        let to = wallet1.address();
+        let data = b"hello";
+        let tx = TransactionRequest::new()
+            .from(from)
+            .to(to)
+            .nonce(3)
+            .value(1000)
+            .data(data)
+            .gas(500_000)
+            .gas_price(1234);
+        let tx_rlp = tx.rlp(chain_id);
+        let sighash = keccak256(tx_rlp.as_ref()).into();
+        let sig = wallet0.sign_hash(sighash, true);
+        let to = tx.to.map(|to| match to {
+            NameOrAddress::Address(a) => a,
+            _ => unreachable!(),
+        });
+        Transaction {
+            from: tx.from.unwrap(),
+            to,
+            gas_limit: tx.gas.unwrap(),
+            gas_price: tx.gas_price.unwrap(),
+            value: tx.value.unwrap(),
+            call_data: tx.data.unwrap(),
+            nonce: tx.nonce.unwrap(),
+            v: sig.v,
+            r: sig.r,
+            s: sig.s,
+            ..Transaction::default()
+        }
+    }
+
+    /// Compute the raw_public_inputs column from the verifier's perspective.
+    fn raw_public_inputs_col<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize>(
```

Comment:
I think we could move this function outside of the test mod and use it to set the witness.
In particular, for the TxTable and BlockTable we have functions to assign the values already defined in the evm circuit.  See:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4bdb2a2e70b6137ec0905bede725f5750fe4caf7/zkevm-circuits/src/evm_circuit.rs#L202 and https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4bdb2a2e70b6137ec0905bede725f5750fe4caf7/zkevm-circuits/src/evm_circuit.rs#L316

So I think this code could be simplified by splitting the assignment to tables, and the assignment to the rest of columns that don't belong to shared tables.

In my PR of the SuperCircuit I've already refactored the table assignment functions so that they can be used from the other circuits.  See https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/d8a0d122b8731885300933b9b1963c2974a4c7ce/zkevm-circuits/src/table.rs#L115

I want to say that there are some details on the approach that I describe that are a bit unclear to me: in the assign functions I shared, each table is assigned in a new region; but for the PublicInputs circuit we want everything at an absolute offset.  So maybe the solution to use those functions here is not straight forward :thinking:

Reply:
I'm not sure how we could reuse `raw_public_inputs_col` to set the witness, but I think reusing the `load..` functions is a good idea. 
In particular  `load_block` works perfectly; the fields and order are the same and as the block table is located at the beginning of the PI circuit there are no issues with the offset.
However, for the `load_txs()` I have noticed that the layout is not quite the same.  When assigning several txs `load_txs()` assigns the values of each tx including `CALLDATA` before starting to assign the next tx values.
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/4bdb2a2e70b6137ec0905bede725f5750fe4caf7/zkevm-circuits/src/evm_circuit.rs#L222-L234
In contrast, in the PI circuit tx  `CALLDATA` are stored together at the end of the table.
https://github.com/davidnevadoc/zkevm-circuits/blob/ec5362c816032560dba4614d5b98146c80b0063d/zkevm-circuits/src/pi_circuit.rs#L666-L751
In any case, I think it is a very good idea to have common functions to do all these assignments. Maybe we should do this unification effort as a separate improvement. WDYT?